### PR TITLE
[build]  Add support for specifying binary links in buildspec

### PR DESCRIPTION
**Pull request by nskool**
*Issue #, if available:*

## Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [build] | [test] | [build, test] | [ec2, ecs, eks, sagemaker]
- [x] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

*Description:*
1. Add functionality to specify a binary link in buildspec. A VAR_IN_DOCKERFILE ref can be specified. This allows the binary to be referred by a particular ARG in the dockerfile.
2. A test checks if the binary is public/private since the binary should be public before PR is merged.

*Tests run:*
Build command used to test:
```
python src/main.py --buildspec tensorflow/buildspec.yml --framework tensorflow --image_types training --device_types cpu --py_versions py3
```
*DLC image/dockerfile:*
Tensorflow 2.1.1 dockerfiles.

*Additional context:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/utils.py [hunk 1]
```diff
@@ -59,6 +60,48 @@ def do_build_all_images():
...
+    Fetch remote files and save with provided local_path name
+    :param link_type: string
+    :param remote_url: string
+    :return: file_name: string
+    """
+    LOGGER.info(f"Downloading {remote_url}")
+
+    ctx = Context()
+    file_name = ctx.run(f"basename {remote_url}").stdout.strip()
+    LOGGER.info(f"basename: {file_name}")
```
Comment (saimidu) [inline src/utils.py:91]: Why does file_name need to be identified through `ctx.run`? Why not use os.path.basename()?

File: src/utils.py [hunk 2]
```diff
@@ -59,6 +60,48 @@ def do_build_all_images():
...
+
+    if link_type in ["s3"] and remote_url.startswith("s3://"):
+        match = re.match(r's3://(.+?)/(.+)', remote_url)
+        bucket_name = match.group(1)
```
Comment (saimidu) [inline src/utils.py:95]: Add escape characters to "/"s in the regular expression.

Comment (saimidu) [reply src/utils.py:95]: Also, how is `(.+?)` different from `(.+)`?

File: src/utils.py (hunk 1 above)
Comment (nskool) [reply src/utils.py]: Thanks, corrected to using os.path.basename.

File: src/utils.py (hunk 2 above)
Comment (nskool) [reply src/utils.py]: Forward slashes needn't be escaped. IDE also recommends it as redundant. However, for generic representation in line with other languages, I've added them back. 

(.+?) will match up to one entry between slashes, that'll form the 'bucket', the rest becomes the key, including slashes.

Comment (saimidu) [reply src/utils.py]: Thanks. I saw recommendations to escape forward slashes, but if it is redundant, then that's fine.

Sure. `(.+?)` will match with a bucket name, but why do we need the `?`? Doesn't it signify that a particular character/group is optional? (it needs to occur between 0 and 1 times)

Comment (nskool) [reply src/utils.py]: In this case - the bucket name is expected to exist and will match up to 1 occurrences of the bucket name.

**Pull request merged**